### PR TITLE
Skip opening irrelevant class files

### PR DESCRIPTION
**Pull request by singlethink**
If we're just going to ignore them, we don't need to open a zip reader
for them or pull a buffer from the pool.

```
name              old time/op    new time/op    delta
Parse-12            64.9µs ± 4%    34.5µs ± 9%  -46.80%  (p=0.008 n=5+5)
ParseParallel-12    6.79µs ± 1%    3.10µs ± 2%  -54.37%  (p=0.008 n=5+5)

name              old alloc/op   new alloc/op   delta
Parse-12            18.1kB ± 0%     4.8kB ± 0%  -73.48%  (p=0.008 n=5+5)
ParseParallel-12    18.6kB ± 0%     5.0kB ± 0%  -73.21%  (p=0.008 n=5+5)

name              old allocs/op  new allocs/op  delta
Parse-12              54.0 ± 0%      30.0 ± 0%  -44.44%  (p=0.008 n=5+5)
ParseParallel-12      54.0 ± 0%      30.0 ± 0%  -44.44%  (p=0.008 n=5+5)
```

### CODE REVIEW

File: jar/jar.go [hunk 1]
```diff
@@ -299,6 +307,14 @@ func (c *checker) checkJAR(r *zip.Reader, depth int, size int64) error {
 	return err
 }
 
+func (c *checker) checkForOldJndiManagerConstructor(filename string) bool {
+	return !c.hasOldJndiManagerConstructor && strings.Contains(filename, "JndiManager")
+}
+
+func (c *checker) checkJndiManagerVersion(filename string) bool {
```
Comment (ericchiang) [inline jar/jar.go:314]: it's weird to have a function that takes a receiver and doesn't use it.

**Review by ericchiang — COMMENTED** [review]
I find this logic really hard to read (which may be our fault).

There's already c.bad(). Can that be  to something like

```go
func (c *checker) shouldCheckClass(filename string) bool {
    if c.bad() {
        // Already determined that the content is bad.
        return false
    }
}
```

File: jar/jar.go (hunk 1 above)
Comment (singlethink) [reply jar/jar.go]: Fixed.

File: jar/jar.go [hunk 2]
```diff
@@ -205,6 +205,14 @@ func (c *checker) checkJAR(r *zip.Reader, depth int, size int64) error {
 				// need to check more.
 				return nil
 			}
+			if !c.hasLookupClass && strings.Contains(p, "JndiLookup.class") {
```
[the remark this answers is not in the document]
Comment (singlethink) [reply jar/jar.go:218]: Agreed.  The existing detection logic is very difficult to follow.  I've refactored it to improve readability and added comments.  In the process, it made it possible to remove the redundant checks without fear that checkClass would be called on an inappropriate class.

File: jar/jar.go [hunk 3]
```diff
@@ -364,11 +380,8 @@ func init() {
 }
 
 func (c *checker) checkClass(filename string, r io.Reader, buf []byte) error {
-	if !c.hasLookupClass && strings.Contains(filename, "JndiLookup.class") {
-		c.hasLookupClass = true
-	}
-	checkForOldJndiManagerConstructor := !c.hasOldJndiManagerConstructor && strings.Contains(filename, "JndiManager")
-	checkJndiManagerVersion := strings.Contains(filename, "JndiManager.class")
+	checkForOldJndiManagerConstructor := c.checkForOldJndiManagerConstructor(filename)
```
[the remark this answers is not in the document]
Comment (singlethink) [reply jar/jar.go]: Solved via refactor.

**Comment by singlethink** [comment]
Attempted to resolve your concerns via refactor.  In the process, I identified some false negatives and one false positive.  I've resolved the false positive and documented the false negatives.

**Review by ericchiang — APPROVED** [review]
This is so much better, thank you for that!

Please squash and I'll merge

**Comment by singlethink** [comment]
Done

**Pull request merged**